activerecord/lib/active_record/relation/calculations.rb

```diff
@@ -360,7 +360,11 @@ def perform_calculation(operation, column_name)
             if !distinct
               distinct = distinct_select?(select_for_count) if group_values.empty?
             elsif group_values.any? || select_values.empty? && order_values.empty?
-              column_name = primary_key
+              column_name = if klass.composite_primary_key?
+                primary_key.join(", ")
```

Comment:
If this isn't supported by anything but mysql we don't have any guards for that. We could implement a `supports_distinct_with_composite_key?` and default adapters to `false` except for mysql. The better option would be to figure out how to support this for all adapters.

Reply:
Just to elaborate on the unsupported syntax a little bit more for better context:
So the unsupported syntax is not specific to the composite primary key setup itself but rather to a `COUNT()` function along with `DISTINCT` by multiple columns.

For example (non-cpk model):
```ruby
Topic.select(:id, :title).distinct.count
```

Produces:
```sql
SELECT COUNT(DISTINCT id, title) FROM "topics"
```

Which results in `wrong number of arguments to function COUNT()` on sqlite and raises a similar error on postgresql.

When mysql2 adapter is used, the produced query looks like:
(the same query but with a different quoting)
```sql
SELECT COUNT(DISTINCT id, title) FROM `topics`
```
and succeeds

So what we could do is to introduce a guard clause like `supports_distinct_count_by_multiple_columns?` and raise if the application attempts to use it. However since it's not composite primary key specific, I would prefer introducing the guard in a separate PR since the change lives out of the CPK context.

#### Potential solution

There is something that could be done to allow `select(:col1, :col2).distinct.count` work with any database. 
For example if we pass the `:all` argument to `Topic.select(:id, :title).distinct.count(:all)` it produces a query with subquery:

```sql
# mysql
SELECT COUNT(*) FROM (SELECT DISTINCT `topics`.`id`, `topics`.`title` FROM `topics`) subquery_for_count

# sqlite

SELECT COUNT(*) FROM (SELECT DISTINCT "topics"."id", "topics"."title" FROM "topics") subquery_for_count
```

And succeeds in both cases. So even though it's a bit more complex query, it is a working query which is much better than an exception. So what we could do is to make the `.distinct.count` usage smart enough to either use `COUNT ( DISTINCT (columns, list) )` on mysql or fallback to a subquery version on any other adapter, including the composite primary key case like `COUNT ( DISTINCT (<composite, primary, key, columns>) )`

Though similar to the guard clause, since the root problem exists in the current implementation of `count` and not directly related to the composite primary key capabilities I would say we should definitely separate it from the current PR and potentially allow this PR to be merged with an expectation of extending it once the root issue with `count` is fixed.

As a workaround, non-mysql applications can execute a similar query using a bit verbose syntax like 
```ruby
Cpk::Book.select(Cpk::Book.primary_key).distinct.count(:all)
```

Reply:
Will do!